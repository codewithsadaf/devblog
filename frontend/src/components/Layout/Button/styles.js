//imports
import styled, { css } from "styled-components";

const handleButtonVariant = {
    default: (theme) => css`
        background: ${theme.colors.primaryColor};
        color: ${theme.colors.white};
        border: 2px solid ${theme.colors.primaryColor};
        border-radius: 4px;
        transition: ${theme.transition.tran_03};

        &:hover {
            background: transparent;
            color: ${theme.colors.primaryColor};
            border: 2px solid ${theme.colors.primaryColor};
            transition: ${theme.transition.tran_03};
        }
    `,
    outline: (theme) => css`
        background: transparent;
        color: ${theme.colors.primaryColor};
        border: 2px solid ${theme.colors.primaryColor};
        border-radius: 4px;
        transition: ${theme.transition.tran_03};

        &:hover {
            background: ${theme.colors.primaryColor};
            color: ${theme.colors.white};
            border: 2px solid ${theme.colors.primaryColor};
            transition: ${theme.transition.tran_03};
        }
    `,

    success: (theme) => css`
        background: ${theme.colors.success};
        color: ${theme.colors.white};
        border: 2px solid ${theme.colors.success};
        border-radius: 4px;
        transition: ${theme.transition.tran_03};

        &:hover {
            opacity: 0.8;
            transition: ${theme.transition.tran_03};
        }
    `,

    danger: (theme) => css`
        background: ${theme.colors.danger};
        color: ${theme.colors.white};
        border: 2px solid ${theme.colors.danger};
        border-radius: 4px;
        transition: ${theme.transition.tran_03};

        &:hover {
            opacity: 0.8;
            transition: ${theme.transition.tran_03};
        }
    `,
};

const handleUppercase = () => css`
    text-transform: uppercase;
`;

const handleMaxWidth = () => css`
    width: 100%;
    padding: 1em 0;
`;

const handleSmallStyles = (theme) => css`
    padding: 7px 30px;
    border-radius: 4px;
    font-size: ${theme.font.sizes.small};
    border: 1px solid;

    &:hover {
        border: 1px solid;
    }
`;

export const ButtonContainer = styled.button`
    ${({ theme, buttonVariant, uppercase, maxWidth, isSmall }) => css`
        padding: 0.5em 1.8em;
        font-size: 1em;
        border: none;
        font-weight: bold;
        cursor: pointer;
        ${handleButtonVariant[buttonVariant](theme)};
        ${uppercase && handleUppercase()};
        ${maxWidth && handleMaxWidth()};
        ${isSmall && handleSmallStyles(theme)};
    `}
`;
