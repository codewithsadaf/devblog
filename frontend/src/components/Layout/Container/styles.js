//imports
import styled from "styled-components";

export const Container = styled.div`
    min-height: 60vh;
    margin-bottom: 5em;
    padding: 0 2em;
`;
