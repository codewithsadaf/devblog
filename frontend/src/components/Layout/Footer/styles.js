//imports
import styled, { css } from "styled-components";

export const FooterWrapper = styled.footer`
    ${({ theme }) => css`
        height: 250px;
        width: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;
        background-color: ${theme.colors.footerBackground};
        text-align: center;
    `}
`;
export const FooterHeading = styled.h3`
    ${({ theme }) => css`
        @media ${theme.media.lteMedium} {
            font-size: ${theme.font.sizes.medium};
            line-height: ${theme.font.line_height.h6};
        }
    `}
`;

export const CopyrightText = styled.span`
    display: block;
    margin-top: 0.8em;
`;
