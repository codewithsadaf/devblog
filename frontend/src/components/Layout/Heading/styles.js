//imports
import styled, { css } from "styled-components";

const headingSize = {
    huge: (theme) => css`
        font-size: ${theme.font.sizes.xhuge};
        line-height: ${theme.font.line_height.h1};
        ${mediaFont(theme)};
    `,
    big: (theme) => css`
        font-size: ${theme.font.sizes.huge};
        line-height: ${theme.font.line_height.h2};
    `,
    large: (theme) => css`
        font-size: ${theme.font.sizes.xxlarge};
        line-height: ${theme.font.line_height.h3};
    `,
    medium: (theme) => css`
        font-size: ${theme.font.sizes.medium};
        line-height: ${theme.font.line_height.h6};
    `,
    small: (theme) => css`
        font-size: ${theme.font.sizes.small};
        line-height: ${theme.font.line_height.small};
    `,
    xsmall: (theme) => css`
        font-size: ${theme.font.sizes.xsmall};
        line-height: ${theme.font.line_height.small};
    `,
};

const mediaFont = (theme) => css`
    @media ${theme.media.lteMedium} {
        font-size: ${theme.font.sizes.xlarge};
        line-height: ${theme.font.line_height.h4};
    }
`;

export const Container = styled.h1`
    ${({ theme, size, uppercase, bold }) => css`
        text-transform: ${uppercase ? "uppercase" : "none"};
        font-weight: ${bold ? "bold" : 500};
        ${headingSize[size](theme)};
    `}
`;
