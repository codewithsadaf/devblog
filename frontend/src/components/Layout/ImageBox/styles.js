//imports
import styled, { css } from "styled-components";

const handleBigStyle = () => css`
    width: 100%;
    height: auto;
`;

export const Container = styled.div`
    ${({ theme }) => css``}
`;

export const ImageElm = styled.img`
    ${({ rounded }) => css`
        width: 150px;
        height: 150px;
        object-fit: cover;
        display: block;
        margin: 1em auto;
        border-radius: ${rounded ? "50%" : "7px"};
    `}
`;
