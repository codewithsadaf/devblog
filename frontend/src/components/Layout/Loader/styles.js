//imports
import styled, { css } from "styled-components";

export const Container = styled.div`
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100vh;
    background: rgba(0, 0, 0, 0.6);
    z-index: 99999;
    pointer-events: all;
`;

export const Circle = styled.div`
    ${({ theme }) => css`
        &,
        &::before,
        &::after {
            border-radius: 50%;
            width: 2.5em;
            height: 2.5em;
            animation-fill-mode: both;
            animation: bblFadInOut 1.8s infinite ease-in-out;
        }

        & {
            color: ${theme.colors.white};
            font-size: 7px;
            position: relative;
            text-indent: -9999em;
            transform: translateZ(0);
            animation-delay: -0.16s;
        }

        &::before,
        &:after {
            content: "";
            position: absolute;
            top: 0;
        }

        &::before {
            left: -3.5em;
            animation-delay: -0.32s;
        }

        &::after {
            left: 3.5em;
        }

        @keyframes bblFadInOut {
            0%,
            80%,
            100% {
                box-shadow: 0 2.5em 0 -1.3em;
            }
            40% {
                box-shadow: 0 2.5em 0 0;
            }
        }
    `}
`;
