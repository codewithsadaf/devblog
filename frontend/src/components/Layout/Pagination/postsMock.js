export const postsMock = [
    {
        _id: "1",
        title: "title 01",
        content: "content 01",
        image: "img.src",
        tags: ["tag1", "tag2"],
        user: {
            name: "john doe",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
    {
        _id: "2",
        title: "title 02",
        content: "content 02",
        image: "img.src",
        tags: ["tag1", "tag2"],
        user: {
            name: "john doe",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
    {
        _id: "3",
        title: "title 03",
        content: "content 03",
        image: "img.src",
        tags: ["tag1", "tag2"],
        user: {
            name: "john doe",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
    {
        _id: "4",
        title: "title 04",
        content: "content 04",
        image: "img.src",
        tags: ["tag1", "tag2"],
        user: {
            name: "john doe",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
];
