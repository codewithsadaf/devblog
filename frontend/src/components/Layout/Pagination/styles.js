//imports
import styled, { css } from "styled-components";

export const PostsContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    width: 100%;
`;

export const PaginationWrapper = styled.div`
    display: flex;
    align-items: center;
    width: 100%;
    justify-content: flex-start;
    margin-top: 1em;
`;

export const PaginationPage = styled.div`
    ${({ theme }) => css`
        padding: 7px;
        height: 40px;
        width: 40px;
        border-radius: 4px;
        text-align: center;
        display: flex;
        align-items: center;
        justify-content: center;
        font-size: 1.2em;
        transition: ${theme.transition.tran_03};
        cursor: pointer;
        margin-right: 0.3em;
        background: transparent;
        color: ${theme.colors.primaryColor};
        border: 1px solid ${theme.colors.primaryColor};
        transition: ${theme.transition.tran_03};

        &.active {
            background: ${theme.colors.primaryColor};
            color: ${theme.colors.white};
            cursor: default;
            pointer-events: none;
        }

        &:hover {
            background: ${theme.colors.primaryColor};
            color: ${theme.colors.white};
            transition: ${theme.transition.tran_03};
        }
    `}
`;

export const PaginationNavigate = styled.div`
    ${({ theme }) => css`
        background: ${theme.colors.primaryColor};
        padding: 7px;
        height: 40px;
        width: 40px;
        transition: ${theme.transition.tran_03};
        background: transparent;
        color: ${theme.colors.primaryColor};
        border: 1px solid ${theme.colors.primaryColor};
        border-radius: 4px;
        text-align: center;
        display: flex;
        align-items: center;
        transition: ${theme.transition.tran_03};
        justify-content: center;
        font-size: 30px;
        cursor: pointer;
        transition: ${theme.transition.tran_03};

        &:hover {
            background: ${theme.colors.primaryColor};
            color: ${theme.colors.white};
            transition: ${theme.transition.tran_03};
        }

        &:first-child {
            margin-right: 0.3em;
        }
    `}
`;
