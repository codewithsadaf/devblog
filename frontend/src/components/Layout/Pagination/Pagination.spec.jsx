//imports
import { fireEvent, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
//styles
import { renderTheme } from "../../../styles/render-theme";
//mocks
import { postsMock } from "./postsMock";
//components
import { Pagination } from "./";

//window.scrollTo mock
const spyScrollTo = jest.fn();
Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });

describe("<Pagination />", () => {
    describe("Rendering", () => {
        beforeEach(() => {
            renderTheme(
                <BrowserRouter>
                    <Pagination posts={postsMock} postsPerPage={3} />
                </BrowserRouter>
            );
        });

        it("should render Pagination next button", () => {
            const nextElm = screen.getByLabelText("Next Page");

            expect(nextElm).toBeInTheDocument();
        });

        it("should render Pagination previous button", () => {
            const nextElm = screen.getByLabelText("Next Page");

            fireEvent.click(nextElm);

            const previousElm = screen.getByLabelText("Previous Page");

            expect(previousElm).toBeInTheDocument();
        });

        it("should render Pagination pages buttons", () => {
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            expect(paginationButtons).toHaveLength(2);
        });

        it("should match snapshot", () => {
            const { container } = renderTheme(
                <BrowserRouter>
                    <Pagination posts={postsMock} postsPerPage={3} />
                </BrowserRouter>
            );

            expect(container.firstChild).toMatchSnapshot();
        });
    });
    describe("Functionality", () => {
        beforeEach(() => {
            renderTheme(
                <BrowserRouter>
                    <Pagination posts={postsMock} postsPerPage={3} />
                </BrowserRouter>
            );
        });

        it("should go to next page when clicking next button", () => {
            const nextButton = screen.getByLabelText("Next Page");
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            fireEvent.click(nextButton);

            expect(paginationButtons[0]).not.toHaveClass("active");
            expect(paginationButtons[1]).toHaveClass("active");
        });

        it("should go to previous page when clicking previous button", () => {
            const nextButton = screen.getByLabelText("Next Page");
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            //go to next page
            fireEvent.click(nextButton);

            expect(paginationButtons[0]).not.toHaveClass("active");
            expect(paginationButtons[1]).toHaveClass("active");

            const previousButton = screen.getByLabelText("Previous Page");

            //go to previous page
            fireEvent.click(previousButton);

            expect(paginationButtons[0]).toHaveClass("active");
            expect(paginationButtons[1]).not.toHaveClass("active");
        });

        it("should go to page 2 when clicking second pagination button", () => {
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            expect(paginationButtons[0]).toHaveClass("active");
            expect(paginationButtons[1]).not.toHaveClass("active");

            fireEvent.click(paginationButtons[1]);

            expect(paginationButtons[0]).not.toHaveClass("active");
            expect(paginationButtons[1]).toHaveClass("active");
        });

        it("should scroll user scroll/window to top on next button click", () => {
            const nextButton = screen.getByLabelText("Next Page");

            fireEvent.click(nextButton);

            waitFor(() => {
                expect(spyScrollTo).toHaveBeenCalledWith({
                    top: 0,
                    behavior: "smooth",
                });
            });
        });

        it("should scroll user scroll/window to top on previous button click", () => {
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            fireEvent.click(paginationButtons[1]);

            const previousButton = screen.getByLabelText("Previous Page");

            fireEvent.click(previousButton);

            waitFor(() => {
                expect(spyScrollTo).toHaveBeenCalledWith({
                    top: 0,
                    behavior: "smooth",
                });
            });
        });

        it("should scroll user scroll/window to top on pagination button click", () => {
            const paginationButtons =
                screen.getAllByLabelText("Pagination Button");

            fireEvent.click(paginationButtons[1]);

            waitFor(() => {
                expect(spyScrollTo).toHaveBeenCalledWith({
                    top: 0,
                    behavior: "smooth",
                });
            });
        });

        it("should hide previous button when user current page is 1", () => {
            const previousButton = screen.queryByLabelText("Previous Page");

            waitFor(() => {
                expect(previousButton).not.toBeInTheDocument();
            });
        });

        it("should hide next button when user current page is equal to total pages", () => {
            const nextButton = screen.queryByLabelText("Next Page");

            //go to next page
            fireEvent.click(nextButton);

            waitFor(() => {
                expect(nextButton).not.toBeInTheDocument();
            });
        });

        it("should render posts correctly in current page", () => {
            let heading = screen.getAllByRole("heading");
            const nextButton = screen.getByLabelText("Next Page");

            expect(heading).toHaveLength(3);

            //go to next page
            fireEvent.click(nextButton);

            heading = screen.getAllByRole("heading");

            expect(heading).toHaveLength(1);
        });
    });
});
