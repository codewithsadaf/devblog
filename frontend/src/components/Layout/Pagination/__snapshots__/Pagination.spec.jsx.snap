// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Rendering should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c6 {
  font-size: 1em;
  line-height: 150%;
}

.c4 {
  text-transform: none;
  font-weight: bold;
  font-size: 2em;
  line-height: 110%;
}

.c1 {
  width: 100%;
  margin: 0 auto 3em;
  max-width: 600px;
}

.c1 .c5 {
  color: #444;
  font-size: 0.8em;
  line-height: 150%;
  font-style: italic;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 1em;
  border-radius: 4px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.8em 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c8 {
  display: block;
  margin-right: 0.5em;
  font-weight: 500;
}

.c8::before {
  content: "#";
  font-weight: bold;
}

.c9 {
  padding: 0.5em 1.8em;
  font-size: 1em;
  border: none;
  font-weight: bold;
  cursor: pointer;
  background: transparent;
  color: #000;
  border: 2px solid #000;
  border-radius: 4px;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c9:hover {
  background: #000;
  color: #fff;
  border: 2px solid #000;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

@media (max-width:768px) {
  .c1 .c3 {
    font-size: 1.7em;
    line-height: 120%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="title 01"
      class="c2"
      src="img.src"
    />
    <h3
      class="c3 c4"
    >
      title 01
    </h3>
    <p
      class="c5 c6"
    >
      Por: 
      john doe
       - 
      5 de Abril 2023
    </p>
    <div
      class="c7"
    >
      <span
        class="c8"
      >
        tag1
      </span>
      <span
        class="c8"
      >
        tag2
      </span>
    </div>
    <button
      class="c9"
      type="button"
    >
      Ler
    </button>
  </div>
  <div
    class="c1"
  >
    <img
      alt="title 02"
      class="c2"
      src="img.src"
    />
    <h3
      class="c3 c4"
    >
      title 02
    </h3>
    <p
      class="c5 c6"
    >
      Por: 
      john doe
       - 
      5 de Abril 2023
    </p>
    <div
      class="c7"
    >
      <span
        class="c8"
      >
        tag1
      </span>
      <span
        class="c8"
      >
        tag2
      </span>
    </div>
    <button
      class="c9"
      type="button"
    >
      Ler
    </button>
  </div>
  <div
    class="c1"
  >
    <img
      alt="title 03"
      class="c2"
      src="img.src"
    />
    <h3
      class="c3 c4"
    >
      title 03
    </h3>
    <p
      class="c5 c6"
    >
      Por: 
      john doe
       - 
      5 de Abril 2023
    </p>
    <div
      class="c7"
    >
      <span
        class="c8"
      >
        tag1
      </span>
      <span
        class="c8"
      >
        tag2
      </span>
    </div>
    <button
      class="c9"
      type="button"
    >
      Ler
    </button>
  </div>
</div>
`;
